kcwitools/extract_optimal.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import print_function
+
+import numpy as np
+from astropy.io import fits
+import matplotlib.pyplot as plt
+from kcwitools import optimal_extract_utility as r
+from linetools.spectra.xspectrum1d import XSpectrum1D
+
+
+def extract_optimal_1D(fluxcube, varcube,header,wavegrid,x1,x2,y1,y2,outfile=None,verbose=False,plot=False):
+	"""
+
+	Args:
+		fluxcube: 3D kcwi fluxcube
+		varcube: 3D KCWI variance cube
+		header:  header information
+		wavegrid:  1D kcwi wavelength grid
+		
+		x1,x2,y1,y2: Four corners of an aribaty rectangular pseudo-slit for extraction
+		outfile: str, optional
+		  FITS file name
+		verbose [False] : optional, prints information on terminal
+		plot [False] : optional plots diagnostic stuff
+	
+	Returns:
+		xspec: XSpectrum1D
+
+	"""	
+
+	# CHECK CHECK CHECK
+	#THESE VALUES NEED TO BE DOUBLE CHECKED. WHICH ONE SHOULD I USE?
+	gain=header['GAIN2']
+	rdnoise=3.5  #  Don't know this value. Please instruct what should I read in here?
+
+
+	# Hardcoding some values
+	sigma=5  # Sigma Clipping Value
+	fit_order=200   # Chebyshev polinomial order
+
+
+	# Create slices within wavelength ranges that are good for KCWI
+	# From now on only deal with these smaller slices
+	slices=np.where((wavegrid >=3500.) & (wavegrid <= 5500.))
+	flux=fluxcube[slices[0],:,:]
+	wave=wavegrid[slices[0]]
+	var_origin=(varcube[slices[0],:,:])
+
+
+
+	# Get the required shapes
+	wavedim, ydim, xdim = flux.shape
+
+
+	# Plot and show the slit position on the white light image
+	if plot==True:
+		whitelight=np.sum(flux, axis=0)
+		plt.figure()
+		plt.imshow(whitelight,cmap=plt.get_cmap('viridis'),origin="lower")
+		plt.colorbar()
+
+		# Showing the slit position
+		plt.plot([y1,y1,y2,y2,y1],[x1,x2,x2,x1,x1],'g')
+		plt.show()
+
+	# create a mask
+	indx1=mask_index(xdim,y1,xdim,0)
+	indx2=mask_index(xdim,y2,xdim,0)
+	indy1=mask_index(ydim,x1,ydim,0)
+	indy2=mask_index(ydim,x2,ydim,0)
+
+	#Making a 3d masked cube, rather than just a masked slice
+	#True=Good, False=Bad
+
+	mask = np.zeros((wavedim, ydim, xdim))
+	mask[:,indy1:indy2,indx1:indx2]=1
+
+	# Create a boolean 3D True False mask
+	aperture_mask=np.isin(mask, 1)
+	# This is the masked cube that can be used for extraction
+	cube=flux*mask
+	var_origin[~aperture_mask]=np.inf
+
+	#Get the indices of the aperture spaxels at each wavelength
+	aperture_indices = np.where((aperture_mask[900,:,:] == True))
+
+
+	N=len(aperture_indices[0])
+
+
+	# Do a boxcar extraction
+	spec=r.extract1d(cube)
+	var=var_origin+rdnoise**2
```

Comment:
Are you sure the `var` cube doesn't already contain the rdnoise?
It ought to.  @astronomeara can tell you.

Reply:
Good question. @astronomeara : please let me know. Depending on the answer, I will modify the code to parse the variance arrays correctly.

Reply:
I see the following in the headers: 
BIASRN1 =              2.77887 / amp1 RN in e- from bias                        
BIASRN2 =              2.77395 / amp2 RN in e- from bias                        
BIASSUB =                    T / bias subtracted?

Reply:
And that's in the variance cube header

Reply:
Hmm. @astronomeara : Does that mean that the header input going in now should be the header to the variance cube? I guess it should have all the information right?  
Changing the code to reflect this.  Does that mean my line var =var +rdnoise**2 is still true?

Reply:
I tweaked the code now and now if bias subtraction is done previously, set read noise= 0.
@profxj : please check that is is not bonkers.